syntax = "proto3";

package iterators;

message TriplePattern {
  string subject = 1;
  string predicate = 2;
  string object = 3;
  string graph = 4;
}

message SavedScanIterator {
  TriplePattern triple = 1;
  uint64 offset = 2;
  int64 cardinality = 3;
}

message SavedProjectionIterator {
  repeated string values = 1;
  oneof source {
    SavedScanIterator scan_source = 2;
    SavedIndexJoinIterator join_source = 3;
    SavedBagUnionIterator union_source = 4;
  }
}

message SavedIndexJoinIterator {
  oneof source {
    SavedScanIterator scan_source = 1;
    SavedIndexJoinIterator join_source = 2;
  }
  TriplePattern inner = 3;
  map<string, string> muc = 4;
  uint64 offset = 5;
}

message SavedBagUnionIterator {
  oneof left {
    SavedProjectionIterator proj_left = 1;
    SavedBagUnionIterator union_left = 2;
  }
  oneof right {
    SavedProjectionIterator proj_right = 3;
    SavedBagUnionIterator union_right = 4;
  }
}

message SavedFilterIterator {
  oneof source {
    SavedProjectionIterator proj_source = 1;
    SavedFilterIterator filter_source = 2;
  }
  string expression = 3;
}

message RootTree {
  oneof source {
    SavedProjectionIterator proj_source = 1;
    SavedBagUnionIterator union_source = 2;
    SavedFilterIterator filter_source = 3;
  }
}
